core/dbt/task/runnable.py

```diff
@@ -79,13 +82,31 @@ def write_manifest(self):
 
     def load_manifest(self):
         self.manifest = ManifestLoader.get_full_manifest(self.config)
+        self._validate_manifest(self.manifest)
         self.write_manifest()
 
+    def _validate_manifest(self, manifest):
+        for node in self.manifest.nodes.values():
+            # TODO this will be modified when we update the internal representation for python model
+            if (
+                node.config
+                and hasattr(node.config, "language")
+                and node.config.language == ModelLanguage.python
+            ):
+                if (
+                    node.config.materialized
+                    not in self.adapter.valid_python_model_materialization()
+                ):
+                    raise_compiler_error(
+                        f"'{node.config.materialized}' is not a supported materialization of python model for the current adapter,\n"
+                        f"current adapter support {self.adapter.valid_python_model_materialization()}",
+                        node,
+                    )
```

Comment:
I like the idea of the decorator! Makes it easy for adapter maintainers to do the right thing, while we still have to count on them to do it. (Fuller conversation about this in https://github.com/dbt-labs/dbt-core/issues/5325#issuecomment-1191804571)

It feels like it would require a _lot_ of mocking to test, which intimidates me, but maybe doesn't intimidate you. My personal approach would be a functional test in `dbt-spark` + `dbt-bigquery` to validate that the Python code shows up in debug-level logs (e.g. using [`run_dbt_and_capture`](https://github.com/dbt-labs/dbt-core/blob/2548ba99369a38fa8c15050e00e4c5f7152d27d6/core/dbt/tests/util.py#L80))

Reply:
I vote for checking the logs too. Not a perfect option, but lots of mocking means that what you're testing isn't very real anyway.

Reply:
Sounds good! I will get the check for materialization into a separate PR, do some test in the adapters(as we don't really have a python job runner here, maybe we should switch out test db to duckdb sometime haha).

Reply:
Can duckdb run python models?

Reply:
Not yet, we can add it and it is a embedded database, which means we are going to execute user code(given user here is just our devs I think that's fine)